trace/rate_limit.go

```diff
@@ -0,0 +1,294 @@
+
+	lock sync.Mutex
+}
+
+func (r *SharedRateLimit) startInterval(start time.Time) {
+	// If we're in the current interval, just reset and keeping going.
+	// We don't get an updated interval that way, but that's OK.
+	printer.Debugln("New sample interval started:", start)
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.SampleIntervalStart = start
+	r.SampleIntervalCount = 0
+	r.SampleIntervalActive = true
+}
+
+func (r *SharedRateLimit) IntervalStarted() bool {
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	return r.SampleIntervalActive
+}
+
+func (r *SharedRateLimit) endInterval(end time.Time) {
+	printer.Debugln("End of sample interval:", end)
+	intervalLength := end.Sub(r.SampleIntervalStart)
+	r.SampleIntervalActive = false
+	r.SampleIntervalCount = 0
+
+	if r.FirstEstimate {
+		r.EstimatedSampleInterval = intervalLength
+		r.FirstEstimate = false
+	} else {
+		alpha := viper.GetFloat64(RateLimitExponentialAlpha)
+		exponentialMovingAverage :=
+			(1-alpha)*float64(r.EstimatedSampleInterval) + alpha*float64(intervalLength)
+		printer.Debugln("New estimate:", exponentialMovingAverage)
+		r.EstimatedSampleInterval = time.Duration(uint64(exponentialMovingAverage))
+	}
+}
+
+func (r *SharedRateLimit) run() {
+	r.epochTicker = time.NewTicker(viper.GetDuration(RateLimitEpochTime))
+	defer r.epochTicker.Stop()
+
+	// Set up the timer so it's non-nil, but it'll immediately fire
+	// (This only resets the counters, in case a packet has already arrived.)
+	r.intervalTimer = time.NewTimer(0)
+	defer r.intervalTimer.Stop()
+
+	// Main loop: handle time events or shutdown
+	for true {
+		select {
+		case <-r.done:
+			return
+		case epochStart := <-r.epochTicker.C:
+			r.startNewEpoch(epochStart)
+		case intervalStart := <-r.intervalTimer.C:
+			r.startInterval(intervalStart)
+		}
+	}
+}
+
+func (r *SharedRateLimit) Stop() {
+	close(r.done)
+}
+
+func (r *SharedRateLimit) startNewEpoch(epochStart time.Time) {
+	printer.Debugln("New collection epoch:", epochStart)
+
+	r.lock.Lock()
+	defer r.lock.Unlock()
+	r.CurrentEpochStart = time.Now()
+
+	// Ensure timer is stopped before reset
+	// I *think* we don't need to drain the channel here (which would require
+	// extra state to do reliably anyway.)
+	r.intervalTimer.Stop()
+
+	// Pick a time for the next sampling interval to start within this epoch.
+	if r.FirstEstimate {
+		// Didn't get a new estimate, just keep collecting everything.
+		r.intervalTimer.Reset(0)
+	} else {
+		upperBound := viper.GetDuration(RateLimitEpochTime) - r.EstimatedSampleInterval
+		randomOffset := time.Duration(rand.Int63n(int64(upperBound)))
+		r.intervalTimer.Reset(randomOffset)
+	}
+
+	// Trigger request expiration for all collectors
+	threshold := epochStart.Add(-1 * viper.GetDuration(RateLimitMaxDuration))
+	for _, child := range r.children {
+		child.epochCh <- threshold
+	}
+}
+
+// Check if request should be sampled; increase the count by one.
+func (r *SharedRateLimit) AllowHTTPRequest() bool {
+	r.lock.Lock()
```

Comment:
Are we concerned that this might become a bottleneck? (If yes, alas, I don't see an obvious design that avoids this.)

Reply:
Somewhat, but I suspect in practice I expect most traffic comes through a single interface so there is not a lot of contention.

I initially thought of doing something more complicated using channels. You can give every collector a channel of size 1 (or, larger than 1), which it must successfully read from before it can record a packet.  It does the read in a nonblocking fashion and drops the packet if it fails.  The share rate limit farms out its budget to any channel that is writeable.  This means the control is not as precise, but that seems acceptable.

The problem with implementing this in Go is that there's no good way to select() on an arbitrary list of channels, as far as I know.  You could do a fan-out using multiple goroutines reading from a single source channel.

A better way is probably batching-- grab N "leases" to send a packet, when the lock is grabbed, instead of just 1, and decrement the count locally per-collector.  If needed, these leases could have an expiration time after which they need to be returned to the pool.

Reply:
Yeah, these options are sufficiently complicated that I think we can just keep them in our back pockets for if/when the bottleneck becomes an issue.